### ISSUE TITLE
Can I use FileReader to process data for target detection or semantic segmentation tasks?

### ISSUE THREAD

**Issue by wangguangyuan** [NVIDIA/DALI#112]
Can I use FileReader to process data for target detection or semantic segmentation tasks? For example, the target detection my data format is image_path [x1, y1, x2, y2 label], or the semantic segmentation task image_path mask_path. 
Is there any other way?

**Comment by wangguangyuan** [comment]
@JanuszL

**Comment by pribalta** [comment]
Hi,

Could you provide more information about the data format you're planning to use for target detection? Do these coordinates refer to bounding boxes? Also, what are the formats of the images?

Similar thing for semantic segmentation. What formats do you plan on using? :)

**Comment by JanuszL** [comment]
Internally tracked at DALI-200

**Comment by ptrendx** [comment]
Hi @wangguangyuan 

That is currently not possible but is a very good suggestion especially as we are currently working on enabling detection workloads (see for example #110). 
As @pribalta mentioned already though, we would like more information on the format of data. For example, bounding boxes may be in the format [x,y,width, height] or [left, right, bottom, top] (and maybe others). Although, to be fair, those are still just 4 numbers so we would not really care.
I guess we could add an additional argument in the `FileReader` that would be basically a format of the file list (with default being the current behavior), what do you think @JanuszL?

**Comment by wangguangyuan** [comment]
I am very sorry, I did not describe the problem clearly.
When I use FileReader to read data, one way is the path to the file passed by file_list. For the classification task, the format of the file is "file label". For target detection tasks, the file should be in the format "file [x,y,width, height] or [left, right, bottom, top]". For semantic segmentation tasks, the format of the file should be "file mask", file is the path of the image, mask is the mask path corresponding to the image, and for the data enhancement of the image, the corresponding mask should be the same data enhancement. 
If FileReader does not implement the above requirements, is there any other way to do this? I saw https://github.com/NVIDIA/DALI/blob/master/dali/benchmark/resnet50_bench.py given the following method, but for semantic segmentation, I still can't make data enhancements to images and masks at the same time.

**Comment by wangguangyuan** [comment]
I saw the COCO dataset reader, a great addition。
Is there a good solution for semantic segmentation?
As an example say I have an image and its segmentation mask. We may want to randomly crop the image, and an identical crop needs to be applied to the mask. This means that the random number(s) generated for the crop must be the same for each (image_i, mask_i) pair in the batch (but different for each (image_j, mask_j) pair.

**Comment by JanuszL** [comment]
@ptrendx it is a good idea, what comes to my mind now it to provide some predefined list of formats we want to support and pass this info as an enum to `FileReader`, one mentioned by @wangguangyuan seems fine (for detection and segmentation).
Do we have any raw data that corresponds to any COCO data set we have?

**Comment by wangguangyuan** [comment]
@JanuszL @ptrendx 
For the target detection task, the image is scaled by the short edge. At this time, the bounding box [x1, y1, x2, y2] we marked should also be scaled in the same scale so that the bounding box can correspond to the scaled image. Dali has a good implementation？
I didn't find a good solution, so I plan to use only dali's FileReader to read images and nvJPEGDecoder to decode images. After using the DALI TensorFlow plugin to get the Ops of TensorFlow, I can use the functions in TensorFlow to simultaneously enhance the data and the bounding box. But I found a problem in the process of doing it. One of the shape parameters of DALIIterator() needs to set the image size, but the image size I input is not fixed, which causes me to not call DALIIterator() correctly, and I can't get the Ops of TensorFlow.
Is there some solution?

**Comment by xyy-Iv** [comment]
@wangguangyuan have you solve the way of importing image enhancement dataset? It is my first time to use this lib. I found it hard to read both original and label images.

**Comment by grez72** [comment]
Hi,

Is there any chance you will implement an option for a custom format for the file_list (with the default being the current behavior), as suggested above?

I'm training a [pytorch model](https://github.com/zhirongw/lemniscate.pytorch) which requires each sample to include (image, label, index), because training relies on the absolute index number of the image.

I attempted to use a custom ExternalSource for this, but the loading was substantially slower (10x).

I also attempted to create a custom op, by copying the FileReader op, but this wasn't straightforward, and I have limited C programming experience.

**Comment by JanuszL** [comment]
Hi,
We still have that in mind, but currently, we are focusing on the most common use cases in our internal roadmap. 
Your request sounds like a good example where we would welcome external contribution.
As some instant solution, I could propose to convert your data to TFRecord format - our TFRecord reader is supposed to read any feature embedded in TFRecord.

**Comment by JanuszL** [comment]
Hi All,

I think you should check the most recent improvements in [the `external_source` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.external_source), especially `parallel` and `prefetch_queue_depth` arguments. This operator should provide a sufficient performance of the data loading while maintaining the coding in python flexibility for the data loading.
